src/Sarif.ValidationTests/DefaultObjectPopulatingVisitor.cs

```diff
@@ -0,0 +1,293 @@
+            {
+                { typeof(bool),    (isRequired) => { return isRequired; } },
+                { typeof(int),     (isRequired) => { return isRequired ? int.MaxValue : 0; } },
+                { typeof(double),  (isRequired) => { return isRequired ? double.MaxValue : 0; } },
+                { typeof(string),  (isRequired) => { return isRequired ? "Required string value." : null; } },
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.contoso.com") : null; } }
+            };
+        }
+
+        // These builders exhaustively populate the generated SARIF file, including non-required JSON primitives
+        public static IDictionary<Type, PrimitiveValueBuilder> GetBuildersForAllPrimitives()
+        {
+            // The values in this output are designed to allow conformance to the schema. In our schema,.
+            // for example, some integer values must be > 0. And so, our optional integer value is 
+            // written as 1.
+            return new Dictionary<Type, PrimitiveValueBuilder>()
+            {
+                { typeof(bool),    (isRequired) => { return true; } },
+                { typeof(int),     (isRequired) => { return isRequired ? int.MaxValue : 1; } },
+                { typeof(double),  (isRequired) => { return isRequired ? double.MaxValue : 1; } },
+                { typeof(string),  (isRequired) => { return isRequired ? "[Required string value]" : "[Optional string value]"; } },
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(1776, 7, 4).ToUniversalTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.value.contoso.com") : new Uri("https://optional.uri.value.contoso.com"); } }
+            };
+        }
+
+        public DefaultObjectPopulatingVisitor(JsonSchema schema, IDictionary<Type, PrimitiveValueBuilder> primitiveValueBuilders = null)
+        {
+            _schema = schema;
+            _typeToPropertyValueConstructorMap = primitiveValueBuilders ?? GetBuildersForRequiredPrimitives();
+        }
+
+        private JsonSchema _schema;
+        private IDictionary<Type, PrimitiveValueBuilder> _typeToPropertyValueConstructorMap;
+
+        // There are two places in the format where object nest references to the defining 
+        // type. 'exception.innerExceptions' is an array of exception objects. 
+        // 'node.children' is another collection of nodes. We need to watch for these
+        // scenarios to prevent re-entrancy and eventual stack consumption when 
+        // producing samples of these types.
+        private int graphNodeParentCount;
+        private int exceptionDataParentCount;
+
+        public override object VisitActual(ISarifNode node)
```

Comment:
Needs a big ol' comment to explain why you're overriding VisitActual and using reflection to access the properties, rather than using the usual VisitXxx overrides and the strongly typed OM.